Sources/MapboxNavigation/ImageDownloader.swift

```diff
@@ -11,47 +11,46 @@ protocol ReentrantImageDownloader {
 }
 
 class ImageDownloader: NSObject, ReentrantImageDownloader, URLSessionDataDelegate {
-    private var sessionConfiguration: URLSessionConfiguration = URLSessionConfiguration.default
+    private let sessionConfiguration: URLSessionConfiguration
 
-    lazy private var urlSession: URLSession = {
-        return URLSession.init(configuration: sessionConfiguration, delegate: self, delegateQueue: nil)
+    private lazy var urlSession: URLSession = {
+        return URLSession(configuration: sessionConfiguration, delegate: self, delegateQueue: nil)
     }()
 
-    private var downloadQueue: OperationQueue
-    private var accessQueue: DispatchQueue
+    private let downloadQueue: OperationQueue
+    private let accessQueue: DispatchQueue
 
-    private var operationType: ImageDownload.Type = ImageDownloadOperation.self
+    private var operationType: ImageDownload.Type
     private var operations: [URL: ImageDownload] = [:]
 
-    private var headers: [String: String] = ["Accept": "image/*;q=0.8"]
+    private let headers: [String: String] = ["Accept": "image/*;q=0.8"]
+
+    init(sessionConfiguration: URLSessionConfiguration = .default,
+         operationType: ImageDownload.Type = ImageDownloadOperation.self) {
+        self.sessionConfiguration = sessionConfiguration
+        self.operationType = operationType
 
-    override init() {
         self.downloadQueue = OperationQueue()
         self.downloadQueue.name = Bundle.mapboxNavigation.bundleIdentifier! + ".ImageDownloader"
-        self.accessQueue = DispatchQueue(label: Bundle.mapboxNavigation.bundleIdentifier! + ".ImageDownloaderInternal", attributes: .concurrent)
-    }
-
-    convenience init(sessionConfiguration: URLSessionConfiguration? = nil, operationType: ImageDownload.Type? = nil) {
-        self.init()
-
-        if let config = sessionConfiguration {
-            self.sessionConfiguration = config
-        }
-        
-        if let op = operationType {
-            self.operationType = op
-        }
+        self.accessQueue = DispatchQueue(label: Bundle.mapboxNavigation.bundleIdentifier! + ".ImageDownloaderInternal")
+        super.init()
+
+        // NOTE: `lazy var` isn't thread safe while `init` methods are.
+        //       `lazy var` for `urlSession` is used to specify `self` as a delegate which only available after
+        //       all properties are initialized ans `super.init` is called if appropriate.
+        //       So we trigger `init` here so that `urlSession` is initialized in a thread safe manner.
+        _ = urlSession
```

Comment:
Does the lazy syntax benefit us here? Maybe it’d be simpler to inline that initialization here.

Reply:
Good questions, thanks for asking. Without going deep into explaining the rationale why I've kept `lazy var` I think we can change `private lazy var urlSession: URLSession` to `private var urlSession: URLSession!`.